### PR TITLE
ARROW-13185: [MATLAB] Create a single MEX gateway function which delegates to specific C++ functions

### PR DESCRIPTION
**Pull request by lafiona**
## Overview

This pull request implements a more scalable and sustainable approach to organizing the C++ functionality that needs to be exposed to MATLAB. It adds a new singular MEX gateway function, `mexfcn('<cpp-function-name>', <cpp-function-argument-1>, ..., <cpp-function-argument-N>)`, which delegates to specific C++ functions. To make use of `mexfcn`, the directory containing `mexfcn.<mex-extension>` must be added to the MATLAB path. 

Advantages to this approach:
- Organization
    - All C++ functions that are exposed to MATLAB are registered in one location
    - Reduce the complexity of managing and linking many MEX files and ensuring that they are added to the MATLAB path
    - Reduce cognitive load for adding new functions
- Avoid polluting the source tree with build artifacts
- Reduce build times, building a single MEX function is faster than building potentially hundreds
- Reduce binary bloat caused by creating a separate MEX file for every MEX function
- Enable flexibility in terms of where C++ implementation files live

## Implementation

1. One MEX function, `mexfcn`, defined in `matlab/src/cpp/arrow/matlab/mex/mex_util.h`, dispatches to individual C++ implementation files. 

For example, to invoke featherread functionality from MATLAB, that is implemented in C++:

``` matlab
>> [variables, metadata] = mexfcn('featherread', 'featherfile.feather');
```

2. Functionality implemented in C++ that we want to expose in MATLAB is registered in a function map in the file `matlab/src/cpp/arrow/matlab/mex/mex_functions.h`.
3. Restructured source tree layout and performed general code clean up in preparation for feature implementation work:
  3.1. Split source code to matlab/src/matlab and matlab/src/cpp
  3.2 Make packages, namespaces, and directories consistent in terms of naming and hierarchy for simplifying navigation and header inclusion.
  3.3 Renamed the MATLAB package name `mlarrow` to `arrow` as the `ml` is superfluous.
4. Refactored `matlab/CMakeLists.txt`:
  4.1. Build shared library, `arrow_matlab`, that contains C++ functionality for the interface.
  4.2. macOS: explicitly add the path of `arrow` to the `rpath` of `arrow_matlab`, as paths of libraries output by imported targets are not automatically included.
  4.3. Windows:
    - Copy shared library, `arrow.dll` to the directory where the MEX function lives.
    - Add the path to MATLAB and GTest shared libraries to the ctest `ENVIRONMENT` when building tests.
    - Specify the release version of MSVC Runtime Libraries for all targets created in the CMake file.
5. Enable the `install` target:
  5.1. Utilize [`CMAKE_INSTALL_PREFIX`](https://cmake.org/cmake/help/v3.0/variable/CMAKE_INSTALL_PREFIX.html) to allow users to customize the install location. The platform-specific default install locations will be used if the user does not specify a custom value.
  5.2. Once installed, the interface's source files and libraries are relocatable, on all platforms.
  5.3. As part of the install step, add the path to the install directory to the [MATLAB Search Path](https://uk.mathworks.com/help/matlab/matlab_env/what-is-the-matlab-search-path.html) by:
     - Option 1: Call `addpath` and `savepath` to modify the `pathdef.m` file that MATLAB uses on startup. This option is on by default. However, it can only be used if CMake has the appropriate permissions to modify `pathdef.m`. This option is toggled on and off by the flag: `MATLAB_ADD_INSTALL_DIR_TO_SEARCH_PATH`.
     - Option 2: Add an `addpath` command to the `startup.m` file located at the [`userpath`](https://uk.mathworks.com/help/matlab/matlab_env/what-is-the-matlab-search-path.html#:~:text=on%20Search%20Path.-,userpath%20Folder%20on%20the%20Search%20Path,-The%20userpath%20folder). This option can be used if a user does not have the permissions to modify the `pathdef.m` file. This option is off, by default, and is toggled on and off by the flag: `MATLAB_ADD_INSTALL_DIR_TO_STARTUP_FILE`.
     - Option 3: Add the path to the install directory to the [`MATLABPATH`](https://uk.mathworks.com/help/matlab/matlab_env/add-folders-to-matlab-search-path-at-startup.html#btpajlw) environment variable. The paths listed in `MATLABPATH` are added to the MATLAB Search Path on start up.
     - The MATLAB Actions install of MATLAB used during CI does not have `userpath` set, nor do we have edit permissions for `pathdef.m`. Therefore, we use the third option and set the `MATLABPATH` environment variable in `matlab.yml`.

## Testing
Qualified `CMakeLists.txt` changes by building and running all tests:
  - On Windows 10 (Ninja and Visual Studio), macOS 11.5 (Make and Ninja), and Debian 10 (Make and Ninja)
  - Configurations: build both Arrow and GTest, use provided `ARROW_HOME`, use provided `GTEST_ROOT`, use both `ARROW_HOME` and `GTEST_ROOT`.

## Future Directions
1. Investigate why the default CMake behavior does not link the test executables against the correct MSVC Runtime libraries (ie. `ucrtbase.dll` versus `ucrtbased.dll`) when building with Ninja on Windows. 
2. Add support for specifying function names and arguments as MATLAB strings to `mexfcn`. Currently, only character vectors are supported. 
3. Refactor `mexfcn` to use [MATLAB Data Arrays](https://uk.mathworks.com/help/matlab/matlab-data-array.html) (MDAs) and [C++ Mex](https://uk.mathworks.com/help/matlab/cpp-mex-file-applications.html).
4. Investigate running MATLAB tests using [`matlab_add_unit_test`](https://cmake.org/cmake/help/latest/module/FindMatlab.html#command:matlab_add_unit_test).

## Notes
1. Thank you for all of your help on this pull request, @sgilmore10 and @kevingurney!

### CODE REVIEW

**Comment by lafiona** [comment]
I apologize for the size of this pull request. We found it tricky to partition our changes into separate pull requests as they involved modifying the build system and refactoring the structure of the source files. However, in the future, our goal is to submit smaller, more incremental pull requests to reflect more isolated changes.

File: matlab/CMakeLists.txt
Comment (kou) [inline matlab/CMakeLists.txt:280]: I don't try this but is this target always run even when GTest files aren't rebuilt?
Can we use `BYPRODUCTS` and `DEPENDS` to prevent needless execution?

File: matlab/CMakeLists.txt
Comment (kou) [inline matlab/CMakeLists.txt:420]: How about using `%PATH%` for GTest too instead of copying GTest files?

File: matlab/CMakeLists.txt
Comment (kou) [inline matlab/CMakeLists.txt:371]: Can we deffer this to a follow-up pull request to reduce changes in this pull request?

**Review by kou — COMMENTED** [review]
Should we add GitHub Actions jobs on macOS and Windows to `.github/workflows/matlab.yml`?

Do you want to work on ARROW-13564 to apply "Co-authored-by" in pull request commits before we merge this?

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: This piece of code is not necessary anymore after adding the path of the shared libraries to the test environment path.

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: This makes sense to me! I have removed the code to create `is_multi_config.txt`. And I will add it in a future pull request for [ARROW-15182](https://issues.apache.org/jira/browse/ARROW-15182).

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: Thanks for this suggestion, it'll be more efficient to skip copying the GTest dlls. I will make this change.

**Comment by lafiona** [comment]
Thank you for your feedback, @kou! 

At this time, `[matlab-actions/setup-matlab]`(https://github.com/matlab-actions/setup-matlab) only supports Linux-based VMs in GitHub Actions. I have created two Jira tickets to track adding GitHub Actions jobs on macOS and Windows once the two platforms are supported: 
- macOS: [ARROW-15195](https://issues.apache.org/jira/browse/ARROW-15195) 
- Windows: [ARROW-15196](https://issues.apache.org/jira/browse/ARROW-15196) 

Thank you for checking in about ARROW-13564, it does not need to block this pull request. Because each of the co-authors pushed commits to this pull request, I think everyone will be represented in the final commit after the pull request is merged. 

I have addressed your code review feedback, please refer to the individual threads for the code specific comments.

File: matlab/CMakeLists.txt
Comment (kou) [inline matlab/CMakeLists.txt:286]: Can we use `%PATH%` for this too?

**Review by kou — COMMENTED** [review]
OK.
Could you update the description of this pull request for the latest code?
We use the description on merge.

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt:286]: Thanks @kou, I've updated the description to reflect the changes that I made in response to your feedback. 

`arrow.dll` is not only used by the tests. On Windows, it needs to be copied to the location of `arrow_matlab.dll` and the MEX function, `mexfcn.mexw64`, so that they can access the Arrow APIs.

Comment (kou) [reply matlab/CMakeLists.txt:286]: In normal use (not test use), users run `cmake --build ... --target install` before they use this module, right?
If it's correct, `arrow.dll` should be located in the same directory as `arrow_matlab.dll` and `mexfcn.mexw64` by `cmake --build ... --target install`.

If your explanation is correct (we need this for not test use too), we should use `install()`  https://cmake.org/cmake/help/latest/command/install.html instead of `add_custom_target()`.

I think that we can use one of them:

  * `install(IMPORTED_RUNTIME_ARTIFACTS)`: https://cmake.org/cmake/help/latest/command/install.html#imported-runtime-artifacts
    * This requires CMake 3.21.
  * install(FILES)`:  https://cmake.org/cmake/help/latest/command/install.html#files

**Comment by edponce** [comment]
Simply out of curiousity, `mexfcn` does not supports C++ functions that take `FunctionOptions` in addition to the function arguments?

**Comment by lafiona** [comment]
@edponce, `mexfcn` only receives input arguments from MATLAB. End users will interact with the MATLAB APIs, which will make use of `mexfcn`. `mexfcn` will dispatch to individual C++ functions that implement the necessary function specific behavior. 

Depending on the needs of a particular C++ function, developers can pass options in the way that they deem appropriate, in order to get that information to the C++ layer. `mexfcn` accepts a variable number of input arguments, of any type, from MATLAB. So a developer could pass an options struct to `mexfcn` in order to propagate any options to the C++ layer. The translation from the options struct to C++ data structures would need to be taken care of by the specific C++ functions.

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: Thank you for this feedback, @kou! I have thought more about how we are structuring the source files and libraries for deployment.

As you mentioned, it is advantageous to use `install()` to move `arrow.dll` to the installed location for normal use.

With the previous behavior of copying to the build tree, the MATLAB tests were able to find `mexfcn` and its dependencies by adding the path to the build tree to the MATLAB Search Path. However, enabling the install step allows a user to install the interface files (including source files and mex functions) to any location that the user specifies. Therefore, we enabled options to add the install directory, specified by `CMAKE_INSTALL_PREFIX`, to the MATLAB Search Path, as part of the install step. I have updated the pull request description with more details to reflect the changes we have made.

For running the C++ tests on Windows and macOS, the paths to `arrow.dll`/`libarrow.dylib` are added to the `ctest` environment to eliminate the need to copy at all.

File: matlab/CMakeLists.txt
Comment (kou) [inline matlab/CMakeLists.txt:103]: Can we always set `IMPORTED_IMPLIB` here?
I expect it's only used on Windows. (It's ignored on non Windows environments such as Linux and macOS.)

File: matlab/CMakeLists.txt
Comment (kou) [inline matlab/CMakeLists.txt:132]: It seems that we can always use `CMAKE_IMPORT_LIBRARY_SUFFIX` here.

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: @kou, thank you for this suggestion. I removed the if-statement to only exclude `.DS_Store` files on macOS, as this logic can be included for Linux and Windows platforms, without affecting the result.

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: This makes sense to me! I've accepted your suggestion to simplify this logic into one line of code.

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: Than you for this suggestion! I will make this change for arrow_shared and the GTest targets.

File: matlab/CMakeLists.txt
Comment (lafiona) [reply matlab/CMakeLists.txt]: Yes, it's true that the `IMPORTED_IMPLIB` variable is only used on Windows. Because imported libraries are not used on macOS and Linux, `CMAKE_IMPORT_LIBRARY_SUFFIX` and `CMAKE_IMPORT_LIBRARY_PREFIX` are not set on these platforms. They are used to create the file path contained by `ARROW_LINK_LIB`. Therefore, the value of `ARROW_LINK_LIB` is not set for macOS and Linux. 

Because of this, I would prefer to only set `IMPORTED_IMPLIB` on Windows.

**Review by kou — APPROVED** [review]
+1

**Comment by lafiona** [comment]
Thank you for code reviewing this change and accepting the pull request, @Kou! 

I am able to see the commit that merged this change into the main branch, however, I see a message on this conversation thread named "Closed with unmerged commits". Do you know what causes this message and are there any action items for me to resolve it?

**Comment by kou** [comment]
@lafiona Sorry for confusing you by our merge method. You don't need extra action for this pull request.

We use our merge script https://github.com/apache/arrow/blob/master/dev/merge_arrow_pr.py to merge a pull request. It squashes commits in the pull request, pushes the squashed commit to master and updates the correspondent JIRA issue. We don't use GitHub's "merge button". So  "Closed with unmerged commits" is seen on GitHub UI. But you don't need extra action because we have changes in this pull request on master.

**Comment by lafiona** [comment]
@Kou, no worries, thank you for explaining why this occurs! Good to know that it is expected behavior. Thanks again for your time reviewing this change!

**Pull request closed without merging**
